src/Lean/Elab/Tactic/ElabTerm.lean

```diff
@@ -385,12 +385,48 @@ private def preprocessPropToDecide (expectedType : Expr) : TermElabM Expr := do
         {indentExpr expectedType}\n\
         is false"
     unless r.isAppOf ``isTrue do
-      throwError "\
-        tactic 'decide' failed for proposition\
-        {indentExpr expectedType}\n\
-        since its 'Decidable' instance reduced to\
-        {indentExpr r}\n\
-        rather than to the 'isTrue' constructor."
+      let msg := MessageData.ofLazyM (es := #[expectedType]) <| withoutModifyingState do
+        -- Re-reduce the instance and collect diagnostics, to get all unfolded Decidable instances
+        withOptions (fun opt => diagnostics.set opt true) do
+          modifyDiag (fun _ => {})
+          discard <| withDefault <| whnf s
+          let unfolded := (← get).diag.unfoldCounter.foldl (init := #[]) fun cs n _ => cs.push n
+          let unfoldedInsts ← unfolded |>.qsort Name.quickLt |>.filterMapM fun n => do
+            let e ← mkConstWithLevelParams n
+            if (← Meta.isClass? (← inferType e)) == ``Decidable then
+              return MessageData.ofConst e
+            else
+              return none
+          let unfoldMsg :=
+            if unfoldedInsts.isEmpty then
+              MessageData.nil
+            else
+              let instances := if unfoldedInsts.size == 1 then "instance" else "instances"
+              m!"after unfolding the {instances}{indentD <| MessageData.andList unfoldedInsts.toList}\n"
+          let hint :=
+            if r.isAppOf ``Eq.rec then
+              m!"\n\n\
+              Hint: reduction got stuck on '▸' (Eq.rec), which indicates that one of the instances \
```

Comment:
I think it would be a little more beginner-friendly to change the wording from "instances" to "decidability instances".